src/shm_main.c

```diff
@@ -768,6 +831,278 @@ sr_shmmain_createunlock(int shm_lock)
+            return NULL;
+        }
+        SR_ERRINFO_SYSERRNO(&err_info, "open");
+        return err_info;
+    }
+
+    /* check the lock */
+    memset(&fl, 0, sizeof fl);
+    fl.l_whence = SEEK_SET;
+    fl.l_start = 0;
+    fl.l_len = 0; /* length of 0 is entire file */
+    fl.l_type = F_WRLCK;
+    rc = fcntl(fd, F_GETLK, &fl);
+    /* Closing any FD to a lock file of a connection owned by this process will
+     * immediately release the lock. When testing locks, we search conn_list
+     * above to ensure we only open/close lock files owned by other processes. */
+    close(fd);
+    if (rc == -1) {
+        SR_ERRINFO_SYSERRNO(&err_info, "flock");
+        return err_info;
+    }
+    if (fl.l_type == F_UNLCK) {
+        SR_LOG_INF("Lock test found no connection for CID %ld.", cid);
+        *conn_alive = 0;
+    } else {
+        /* we cannot get the lock, it must be held by an alive connection */
+        SR_LOG_INF("Lock test found alive connection by pid [%ld] for CID %ld.", fl.l_pid, cid);
+        *conn_alive = 1;
+    }
+    return NULL;
+}
+
+/**
+ * @brief Release the specified connection ID.
+ * Free all resources used in tracking the connection.
+ *
+ * @param[in] cid Connection ID to release.
+ * @return err_info, NULL on success.
+ */
+sr_error_info_t *
+sr_shmmain_deallocate_conn(const sr_cid_t cid)
+{
+    sr_error_info_t *err_info = NULL;
+    char *path = NULL;
+    conn_list_entry *ptr = NULL;
+    conn_list_entry *prev = NULL;
+    conn_list_entry *removed = NULL;
+
+    /* Search the list of active connections owned by this process in order to
+     * remove the connection from the list and clean up */
+    if (conn_list.conn_head) {
+        if ((err_info = sr_mlock(conn_list.conn_list_lock, 1000, __func__))) {
+            return err_info;
+        }
+
+        for (ptr = conn_list.conn_head; ptr;) {
+            if (cid == ptr->cid) {
+                /* remove the entry from the list */
+                if (!prev) {
+                    conn_list.conn_head = ptr->_next;
+                } else {
+                    prev->_next = ptr->_next;
+                }
+                /* cleanup local resources */
+                if (ptr->cid_fd > 0) {
+                    /* Note that closing ANY file descriptor to a locked file
+                     * releases all locks */
+                    close(ptr->cid_fd);
+                    ptr->cid_fd = 0;
+                } else {
+                    SR_LOG_WRN("Connection CID %d fd is unset", ptr->cid);
+                }
+                /* removed entry does not adjust prev */
+                /* prev = prev; */
+                removed = ptr;
+                ptr = ptr->_next;
+                free(removed);
+            } else {
+                prev = ptr;
+                ptr = ptr->_next;
+            }
+        }
+        sr_munlock(conn_list.conn_list_lock);
+    }
+
+    sr_shmmain_get_conn_lockfile_path(cid, &path);
+    if (0 != unlink(path)) {
```

Comment:
Please switch the operands, I am too used to it the other way around that this confuses me.

Reply:
Done